estimagic/optimization/process_arguments.py

```diff
@@ -0,0 +1,311 @@
+
+            # Assert that all values are of the correct type
+            for c in candidate:
+                if not isinstance(c, scalar_type):
+                    raise ValueError(msg)
+
+    # Scalar
+    else:
+        num_opt = 1
+
+        # Assert that scalar is of the correct type
+        if not isinstance(candidate, scalar_type):
+            raise ValueError(msg)
+
+    return num_opt
+
+
+def _get_n_opt_and_check_type_nested_list_argument(candidate, argument_required, name):
+    """Calculate number of inputs and check input types for argument
+    that is expected as list/tuple or nested list/tuple.
+
+    Args:
+        candidate
+            Argument
+
+        argument_required (bool):
+            Whether an empty argument is allowed
+
+        name (str):
+            Name of argument used in error message
+
+    """
+    msg = f"{name} has to be a list/tuple or a nested list/tuple."
+
+    # No list or tuple
+    if not isinstance(candidate, (list, tuple)):
+        raise ValueError(msg)
+
+    # Empty list/tuple
+    elif candidate in ([], ()):
+        num_opt = 1
+        if argument_required:
+            raise ValueError(f"{name} needs to be specified")
+
+    # Nested list/tuple
+    elif isinstance(candidate[0], (list, tuple)):
+        num_opt = len(candidate)
+        for c in candidate:
+            if not isinstance(c, (list, tuple)):
+                raise ValueError(msg)
+
+    # Non-empty 1-dimensional list/tuple
+    else:
+        num_opt = 1
+    return num_opt
+
+
+def broadcast_argument(argument, len_arg, n_opts_total, name):
+    """Broadcast argument if of length 1.
+
+    Args:
+        argument:
+            Argument
+
+        len_arg (int):
+            Length of argument
+
+        n_opts_total (int):
+            Number of optimizatiions that should be run
+
+        name (str):
+            Name of argument used in error message (not used so far)
+    """
+    msg = (
+        f"All arguments entered as list/tuple must be of the same length."
+        + f"The length of {name} is below the length of another argument."
+    )
+
+    # Only one optimization
+    if n_opts_total == 1:
+
+        # Select element if entered as list of length 1
+        if isinstance(argument, (list, tuple)) and len(argument) == 1:
+            res = argument[0]
+        else:
+            res = argument
+
+    # Argument is broadcasted
+    elif len_arg == 1:
+
+        # Select element if entered as list of length 1
+        if isinstance(argument, (list, tuple)) and len(argument) == 1:
+            res = [copy.deepcopy(argument[0]) for i in range(n_opts_total)]
+        else:
+            res = [copy.deepcopy(argument) for i in range(n_opts_total)]
+
+    # Argument was entered as list/tuple
+    elif len_arg == n_opts_total:
+        res = argument
+
+    # Argument was entered as too short list/tuple
+    else:
+        raise ValueError(msg)
+    return res
+
+
+def process_optimization_arguments(
```

Comment:
This function should be at the top of the module because, so you see it first and can then decide if you also want to look at the implementation details.